xclim/indices/run_length.py

```diff
@@ -755,26 +755,133 @@ def extract_events(
+
+    Returns
+    -------
+    xr.DataArray
+        Start of the season, units depend on `coord`.
+
+    See Also
+    --------
+    season
+    season_end
+    season_length
+    """
+    return first_run_before_date(da, window=window, date=mid_date, dim=dim, coord=coord)
+
+
+def season_end(
+    da: xr.DataArray,
+    window: int,
+    mid_date: DayOfYearStr | None = None,
+    dim: str = "time",
+    coord: str | bool | None = False,
+    _beg: xr.DataArray | None = None,
+) -> xr.DataArray:
+    """End of a season.
+
+    See :py:func:`season`. Similar to :py:func:`first_run_after_date` but here a season
+    must have a start for an end to be valid. Also, if no end is found but a start was found
+    the end is set to the last element of the series.
+
+
+    Parameters
+    ----------
+    da : xr.DataArray
+        Input N-dimensional DataArray (boolean).
+    window : int
+        Minimum duration of consecutive values to start and end the season.
+    mid_date : DayOfYearStr, optional
+        The date (in MM-DD format) that a run must include to be considered valid.
+    dim : str
+        Dimension along which to calculate consecutive run (default: 'time').
+    coord : str, optional
+        If not False, the function returns values along `dim` instead of indexes.
+        If `dim` has a datetime dtype, `coord` can also be a str of the name of the
+        DateTimeAccessor object to use (ex: 'dayofyear').
+
+    Returns
+    -------
+    xr.DataArray
+        End of the season, units depend on `coord`.
+        If there is a start is found but no end, the end is set to the last element.
+
+    See Also
+    --------
+    season
+    season_start
+    season_length
+    """
+    # Fast path for `season` and `season_length`
+    if _beg is not None:
+        beg = _beg
+    else:
+        beg = season_start(da, window=window, dim=dim, mid_date=mid_date, coord=False)
+    # Invert the condition and mask all values after beginning
+    # we fillna(0) as so to differentiate series with no runs and all-nan series
+    not_da = (~da).where(da[dim].copy(data=np.arange(da[dim].size)) >= beg.fillna(0))
+    end = first_run_after_date(
+        not_da, window=window, dim=dim, date=mid_date, coord=False
+    )
+    if _beg is None:
+        # Where end is null and beg is not null (valid data, no end detected), put the last index
+        # Don't do this in the fast path, so that the length can use this information
+        end = xr.where(end.isnull() & beg.notnull(), da[dim].size - 1, end)
+        end = end.where(beg.notnull())
+    if coord:
+        crd = da[dim]
+        if isinstance(coord, str):
+            crd = getattr(crd.dt, coord)
+        end = lazy_indexing(crd, end)
+    return end
+
+
 def season(
     da: xr.DataArray,
     window: int,
-    date: DayOfYearStr | None = None,
+    mid_date: DayOfYearStr | None = None,
     dim: str = "time",
     coord: str | bool | None = False,
 ) -> xr.Dataset:
```

Comment:
don't have much to say computation-wise, well done. I like how the fast path simplifies things in `season` (we don't mess with `no_end` now) well done